### PR TITLE
[DNM] *: support push operator with stream

### PR DESCRIPTION
**Pull request by nolouch**
should after https://github.com/pingcap/pd/pull/613 
PTAL @disksing  @siddontang

### CODE REVIEW

File: server/grpc_service.go [hunk 1]
```diff
@@ -243,6 +244,13 @@ func (s *Server) RegionHeartbeat(server pdpb.PD_RegionHeartbeatServer) error {
 		if err != nil {
 			return errors.Trace(err)
 		}
+		if isNewStream {
+			storeID := request.GetLeader().GetStoreId()
+			s.streams.Lock()
+			s.streams.regionHeartbeatStreams[storeID] = server
```
Comment (siddontang) [inline server/grpc_service.go]: How do we do if we also have a stream for the storeID?

File: server/grpc_service.go [hunk 2]
```diff
@@ -243,6 +244,13 @@ func (s *Server) RegionHeartbeat(server pdpb.PD_RegionHeartbeatServer) error {
...
+			s.streams.Unlock()
+			isNewStream = false
+		}
```
Comment (siddontang) [inline server/grpc_service.go]: it is not safe to send stream concurrently, the streams Lock can't guarantee the safety here. You should use a goroutine to send individually.

File: server/coordinator.go [hunk 3]
```diff
@@ -53,6 +56,47 @@ var (
...
+		resp: resp,
+		err:  make(chan error),
+	}
+	for uid, ch := range c.chs {
+		ch <- req
+		select {
+		case err := <-req.err:
+			if err != nil {
+				log.Infof("push region hearbeat response faild: %+v error: %s", resp, err)
+				if strings.Contains(err.Error(), "the stream has been done") {
```
Comment (siddontang) [inline server/coordinator.go]: strange to check message here directly, can we have a better way?

File: server/cluster_worker_test.go [hunk 4]
```diff
@@ -445,9 +454,11 @@ func (s *testClusterWorkerSuite) TestHeartbeatChangePeer(c *C) {
 
 	leaderPeer := s.chooseRegionLeader(c, region)
 	c.Logf("[leaderPeer]:%v, [region]:%v", leaderPeer, region)
+	fmt.Println("Debug 1")
```
Comment (siddontang) [inline server/cluster_worker_test.go]: remove Println

File: server/coordinator.go (hunk 3 above)
Comment (nolouch) [inline server/coordinator.go]: now delete it directly,  the stream will close when we meet error.

File: server/operator.go [hunk 5]
```diff
@@ -225,6 +226,7 @@ type regionOperator struct {
 	Ops    []Operator    `json:"ops"`
 	Kind   ResourceKind  `json:"kind"`
 	State  OperatorState `json:"state"`
+	mu     *sync.RWMutex
```
Comment (siddontang) [inline server/operator.go]: why pointer?

Comment (nolouch) [inline server/operator.go]: Umm. maybe for print more friendly.
the operator String() function  use `fmt.Sprintf`, and golint report 
```
call of fmt.Sprintf copies lock value: server.regionOperator contains sync.RWMutex.
```

**Pull request closed without merging**